flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptorFactory.java

```diff
@@ -428,33 +434,66 @@ public int getIndex() {
-                ShuffleDescriptorAndIndex[] shuffleDescriptors) throws IOException;
+                ShuffleDescriptorAndIndex[] shuffleDescriptors, int numConsumer) throws IOException;
     }
 
     private static class DefaultShuffleDescriptorSerializer implements ShuffleDescriptorSerializer {
         private final JobID jobID;
         private final BlobWriter blobWriter;
+        private final int offloadShuffleDescriptorsThreshold;
 
-        public DefaultShuffleDescriptorSerializer(JobID jobID, BlobWriter blobWriter) {
+        public DefaultShuffleDescriptorSerializer(
+                JobID jobID, BlobWriter blobWriter, int offloadShuffleDescriptorsThreshold) {
             this.jobID = jobID;
             this.blobWriter = blobWriter;
+            this.offloadShuffleDescriptorsThreshold = offloadShuffleDescriptorsThreshold;
         }
 
         public MaybeOffloaded<ShuffleDescriptorAndIndex[]> serializeAndTryOffloadShuffleDescriptor(
-                ShuffleDescriptorAndIndex[] shuffleDescriptors) throws IOException {
+                ShuffleDescriptorAndIndex[] shuffleDescriptors, int numConsumer)
+                throws IOException {
 
             final CompressedSerializedValue<ShuffleDescriptorAndIndex[]> compressedSerializedValue =
                     CompressedSerializedValue.fromObject(shuffleDescriptors);
 
             final Either<SerializedValue<ShuffleDescriptorAndIndex[]>, PermanentBlobKey>
                     serializedValueOrBlobKey =
-                            BlobWriter.tryOffload(compressedSerializedValue, jobID, blobWriter);
+                            shouldOffload(shuffleDescriptors, numConsumer)
+                                    ? BlobWriter.offloadWithException(
+                                            compressedSerializedValue, jobID, blobWriter)
+                                    : Either.Left(compressedSerializedValue);
 
             if (serializedValueOrBlobKey.isLeft()) {
                 return new TaskDeploymentDescriptor.NonOffloaded<>(serializedValueOrBlobKey.left());
             } else {
                 return new TaskDeploymentDescriptor.Offloaded<>(serializedValueOrBlobKey.right());
             }
         }
+
+        /**
+         * Determine whether shuffle descriptors should be offloaded by blob server.
+         *
+         * @param shuffleDescriptorsToSerialize shuffle descriptors to serialize
+         * @param numConsumers how many consumers this serialized shuffle descriptor should be sent
+         * @return whether shuffle descriptors should be offloaded by blob server
+         */
+        private boolean shouldOffload(
+                ShuffleDescriptorAndIndex[] shuffleDescriptorsToSerialize, int numConsumers) {
+            return shuffleDescriptorsToSerialize.length * numConsumers
+                    >= offloadShuffleDescriptorsThreshold;
+        }
     }
+
+    @Experimental
+    @Documentation.ExcludeFromDocumentation(
```

Comment:
I think the `@Documentation.ExcludeFromDocumentation` is not needed here. We can change it to comments.  And we generally put the option at the front of a class.